common/src/main/java/com/wynntils/features/user/overlays/map/MiniMapOverlayFeature.java

```diff
@@ -100,33 +110,131 @@ public void render(PoseStack poseStack, float partialTicks, Window window) {
+
+            // TODO compass icon
+
+            // cursor
+            renderCursor(poseStack, centerX, centerZ);
+
+            // disable mask & render border
+            switch (maskType) {
+                case Rectangular -> {
+                    RenderSystem.disableScissor();
+                    renderRectangularMapBorder(poseStack, renderX, renderY, width, height);
+                }
+                    // case Circle -> {
+                    // TODO
+                    // renderCircularMapBorder();
+                    // }
+            }
+
+            // Directional Text
+            renderCardinalDirections(poseStack, width, height, centerX, centerZ);
+
+            // Coordinates
+            if (showCoords) {
+                String coords = String.format(
+                        "%s, %s, %s",
+                        (int) McUtils.player().getX(), (int) McUtils.player().getY(), (int)
+                                McUtils.player().getZ());
+
+                FontRenderer.getInstance()
+                        .renderText(
+                                poseStack,
+                                centerX,
+                                renderY + height + 10 * height / DEFAULT_SIZE,
+                                new TextRenderTask(
+                                        coords,
+                                        TextRenderSetting.CENTERED.withTextShadow(FontRenderer.TextShadow.OUTLINE)));
+            }
+        }
+
+        private void renderCardinalDirections(
+                PoseStack poseStack, float width, float height, float centerX, float centerZ) {
+            if (showCompass == CompassRenderType.None) return;
+
+            float northDX;
+            float northDY;
 
-            // disable rotation if necessary
             if (followPlayerRotation) {
-                poseStack.popPose();
+                float yawRadians = (float) Math.toRadians(McUtils.player().getYRot());
+                northDX = (float) StrictMath.sin(yawRadians);
+                northDY = (float) StrictMath.cos(yawRadians);
+                if (maskType == MapMaskType.Rectangular) {
+                    // Scale as necessary
+                    double toSquareScaleNorth = Math.min(width / Math.abs(northDX), height / Math.abs(northDY)) / 2;
+                    northDX *= toSquareScaleNorth;
+                    northDY *= toSquareScaleNorth;
+                }
+            } else {
+                northDX = 0;
+                northDY = -height / 2;
             }
 
-            // TODO minimap icons
+            FontRenderer.getInstance()
+                    .renderText(
+                            poseStack,
+                            centerX + northDX,
+                            centerZ + northDY,
+                            new TextRenderTask("N", TextRenderSetting.CENTERED));
 
-            // TODO compass icon
+            if (showCompass == CompassRenderType.North) return;
```

Comment:
Aha, I see, you provide the option to only render "N", so this way it is a slight optimization. Hm. I think readability is more important, but if you feel differently, keep it as it is.

Reply:
I feel as though this is worth it overall, it kind of indicates that `eastDX` and `eastDY` only matter later